### DISCUSSION TITLE
Delay compensation doesn’t work right

### DISCUSSION THREAD

**Discussion by Andi2304** [apohl79/audiogridder#954] [Help]
In Beta8 and Beta9 the delay compensation doesn’t work right on my system. If I put AudioGridder on a track it works, on an aux or master the audio is exactly 516 samples ahead at 44.1k and also at 48k. I am pretty sure Beta7 worked. I am on ProTools, Mojave, Intel plugin side and Windows 10 server side. Best - Andi

**Comment by apohl79** [comment]
Hm sounds strange. The delay compensation code has not changed since many versions, so it might be caused by something else. Can you try different buffer sizes in AG? Can you fix your issue when downgrading to beta8?

**Comment by Andi2304** [comment]
I downgraded to Beta6. That Version works.

Beta8 and Beta9 does not.

I checked different buffer sizes (Beta9):
1 Block (+5ms) 512 samples ahead
2 Blocks (+11ms) 1024 samples ahead
4 Blocks (+23ms) 2048 samples ahead
8 Blocks (+46ms) 4096 samples ahead

Best

Andi

**Reply by apohl79** [reply]
are you using a specific plugin? or is this happening with an empty AG plugin?

**Reply by Centripidity** [reply]
Any idea where I can get hold of beta 6 to try this?  I'm looking for the Mac client and Windows server.

**Reply by Centripidity** [reply]
Ok, found them on AudioGridder.com

**Comment by Andi2304** [comment]
Oh - at Buffer Size Disabled (+0ms) Beta9 works :-)

Best

Andi

**Comment by Andi2304** [comment]
This is also happening with an empty AG plugin.

Best

Andi

**Comment by apohl79** [comment]
Hey Andi,

looks like I need some help reproducing this in PT, as I don't get the audio to be at a wrong offset. Here is what I tried:

- created an audio track with a few clicks
- created an aux track with an empty AG instance with a 2 blocks buffer (1 block is 1024 samples in this setup) => 2048 samples delay
- routed the audio track to the aux 
- committed the aux track and compared the audio data (see screenshot)
- looks all correct to me

<img width="924" alt="image" src="https://user-images.githubusercontent.com/1129160/171136192-21e67844-f873-430f-b1e6-101961e07c5f.png">

**Comment by Andi2304** [comment]
Hey Andreas,

I did the same test. Still ahead when using 1+ Blocks. I use Parallels/Windows10 on the same machine (Mac Pro 2012) by the way.

Beta6 works.

Best

Andi

**Comment by steveschow** [comment]
I'm also getting bad latency compensation also with beta9.  How can I do the workaround to disable the extra buffer?  I'm my case I am using a plugin that has transport sync with the DAW  (Jamstix)...so its always possible the midi sync is the thing that is off...but does not play back in sync with main DAW now...its severely delayed

**Comment by steveschow** [comment]
I tried disabling the buffer, but of course now I get some drop outs.  With any buffer added, playback is not in sync.  

Plugin I am using is Razoon Jamstix (there is a free version to test with).  I don't have any way to verify whether AudioGridder plugin is reporting the extra buffer latency to the DAW...I will assume it is...that seems pretty basic.  But still, the playback is not in sync when using the transport...Jamstix responds to the transport info coming from the host.  Just a guess.

I'm also wondering why in LogicPro, the AGridder.AUinst plugin only displays mono as available, there is no stereo available?  The instrument track in question is a stereo track.

**Comment by steveschow** [comment]
I also had a message flash up in LogicPro at one point complaining that the sample rate was not matching and was 41.xx sample rate...??  I hit ok and it hasn't complained again, but I don't know why it would have complained about the sample rate or whether its ok now or is doing some undesirable internal sample rate conversion.

**Comment by steveschow** [comment]
More specifically...  When I use a single buffer, I end up with +43ms latency reported by audiogridder.  If I bounce the result in my DAW, the audio is 2ms early.

Strangely, when I set audiogridder to 2 buffers, the reported latency is +85ms, but bouncing to an audio track, everything is 71ms early.  

with audiogridder buffer disabled, the audio plays in sync, but occasional dropouts.

**Comment by apohl79** [comment]
Would you be able to provide the plugin and server log files? Cheers! https://audiogridder.com/bug-report-diagnostics-locations/

**Reply by steveschow** [reply]
I will try to test and gather that info this weekend

**Comment by Centripidity** [comment]
I’m having the same issue.  Running Jamstix on a Windows 10 NUC with Logic as the client.

**Comment by steveschow** [comment]
beta6 is not in sync for me either...

**Reply by Centripidity** [reply]
Yeah, I just discovered that.  If I run with 2 buffers the audio seems to be about 12ms ahead of where it should be so it looks like Logic is, for some reason, overcompensating.  Just a passing thought, could it be related to the fact that an FX plugin passes audio in and back but an instrument receives but sends no audio.  Just a wild thought.

I'll do some more testing.

Have you checked if the same problem exists with the 1.1 stable release?

With beta 6 I was unable to play successfully, too much drop out and even apparent sample rate conflicts.  However, I did a short test and was able to bounce successfully with no buffer.

Also, as a work around for no stereo version being available just choose the multi-output version and make sure "Downmix" is switched on in the Jamstix mixer.

**Comment by steveschow** [comment]
I get similar results with DP.  LogicPro is not to blame.

**Comment by steveschow** [comment]
I haven't tried last stable release, yet, but I will

**Comment by steveschow** [comment]
Stable version 1.1.1 is playing in sync.  LogicPro also sees the stereo version.

**Reply by Centripidity** [reply]
Unfortunately, it doesn't have the multi-out version which is what I generally use.  Hopefully this gets fixed soon.

**Comment by steveschow** [comment]
Yea I hear ya.  I tried also to use the AUmfx version of AudioGridder...to host Jamstix as a midi plugin....and it worked...but sync was off a bit with that too.  The midi coming back into LogicPro from audiogridder seems to be live and not using sample accurate timestamps, which makes AudioGridder very unusable for midi plugin hosting of any kind IMHO.  But anyway, using that you can then host Jamstix as a sound module or any other drum plugin back in LogicPro...  Just use Jamstix to generate the midi and return midi instead of audio, as a midi plugin.  it does work, but as I said, the midi was not in sync...and I would say its also jittering around...which to me means AudioGridder is just using a virtual midi port or something to bring the midi back somehow..which loses sample accurate timestamps and causes the jitter.  Just guessing.  Hope that can be fixed too.

Another possibility would then be to host Jamstix inside of a program like Patchworks or Kushview Element.  Then do all the drum mixing there inside that tool (on the audiogridder server).  Less then ideal..but might get the job done.  This is a better way to use Jamstix with a third party drum plugin also in order to get sample accurate timing.

**Comment by Centripidity** [comment]
This is getting a bit weird.  I did a few tests (using beta 9) with all of the different buffer sizes to see how early the audio is arriving by bouncing a short drum sequence and then measuring the displacement of the bounced wave relative to the zero buffer bounce.  These are the results.  I redid the 278ms and 464ms bounces to check that I got the same results.

As you can see, the wave moves further back as the buffer size increases but thenfor some reason the 20, 24 and 28 block buffers are bouncing correctly, **but** when you play the instrument in real time (Jamstix itself, not the bounced track) it is way out of sync.

Very strange behaviour.

<img width="971" alt="Screen Shot 2022-06-30 at 3 34 40 pm" src="https://user-images.githubusercontent.com/108377254/176600947-49c48b56-c8e4-4f4e-89fa-eee37af760b4.png">

**Reply by apohl79** [reply]
thx a mill for this! will try to reproduce and chaise down that issue!

**Reply by Centripidity** [reply]
Happy to help, and if there are other tests you want me to do then please let me know.

I did another test using a 4 block buffer (92ms) and this time the live playback was still out of sync but the bounce was properly synced.

It seems as though, for whatever reason, Logic is playing the audio from AG too soon but, when a bounce is done the results are inconsistent.  Most of the time it bounces the same as realtime playback (ahead of expected time) but sometimes the bounce appears to be in sync.

Also let me say, this is a great piece of software and if you can get it working reliably with low latency and multi-channel audio then it will be extremely useful to me.  Whether commercialised or remaining as donation ware I'd be happy to pay for it.

**Reply by apohl79** [reply]
could you send me your test session logic file/bundle? tried jamstix on windows, beta9, and it's always in perfect sync. I created a midi region and I'm sending a bunch of notes. then I bounced in place with no and a 4 blocks buffer. and I played the bounces live with the instrument. there must be something you guys are doing differently.

**Reply by steveschow** [reply]
Jamstix is meant to be used with its own playback engine, not as a sound module.  It has super smart intelligence about generating drum parts on the fly during playback.

So to properly test it you need to make a test song in jamstix.  When you hit play on the host daw, jamstix plays along the pattern from its own system.  This means jamstix needs to play in sync with the transport of its host daw.

I suspect there is something amiss with audiogridder related to the actual transport sync and the extra buffers that are introduced for improving network reliability.

Additionally, jamstix is often used by some people  in midi mode only where it outputs only midi and may or may not output audio at the same time.  This output midi can then be sent to a better sounding drum plugin such as superior drummer 3, etc  as I mentioned above, when I attempted to do that with suduogridder by using auduogridder plugin from logicpro's midi fx slot; the midi timing of midi events coming back from AG server to logicpro were jittering all over the place, which is probably a separate issue then the transport sync.  It's as if midi from the server is coming back over a virtual midi port as live midi rather then being returned to the host in the process block buffer with sample accurate timestamps.

Two separate but related problems.  Probably

**Reply by steveschow** [reply]
I will try to Install the free version of jamstix today and send you a logicpro project

**Reply by apohl79** [reply]
thx for that explanation!

**Reply by Centripidity** [reply]
I'll put together s simple project to send you.  The projects I've been using to experiment are probably no good to you because they use Jamstix expansions that don't come with the free version.  If I get change I'll do it tonight.

**Reply by Centripidity** [reply]
The problem I have is that I don't know which drummers, styles and kits are available in the free version.  Can anyone tell me so I can create a project that will run successfully?

**Reply by apohl79** [reply]
It's ok, I will try to replicate it using the steps steve was providing. cheers!

**Reply by Centripidity** [reply]
Great, but please let me know if there is anything I can do to help.  Getting this to work would be a major boost to my workflow so I'm keen to see it succeed if possible.

**Reply by apohl79** [reply]
managed to reproduce it with beta 9, and seems like I fixed it. when I follow steve's steps below, the audio comes in in sync now.

**Reply by Centripidity** [reply]
Is the fixed version available?

**Reply by apohl79** [reply]
not yet

**Reply by Centripidity** [reply]
I look forward to its release.

**Comment by steveschow** [comment]
I can send you a logicPro project, but there is really nothing to it...  just a track with Audiogridder on it, and AGserver is hosting the free jamstix.  The key is to create a song in Jamstix Free.  To do this...do the following steps..

1. When Jamstix GUI first comes up, you will see the following:

![Screen Shot 2022-06-30 at 11 46 07 AM](https://user-images.githubusercontent.com/9116800/176744122-65dabd79-7790-4a11-b87d-2321a8230761.jpg)

2. Click on the "Create Song" button.

3. Rewind LogicPro to the first bar if not there already.

4. Hit play.  It will start playing and you should notice that the jamstix transport is also moving, at bar two it will start making drumstick click sounds and at bar 3 it will start playing that actual beat and you will be able to see that as it plays Jamstix is calculating the patterns and adding them to Jamstix's own grid as it plays.

5. After playing it through, you can go back and hit play again and it will play back without changing Jamstix's grid, but its still basically playing back the pattern from Jamstix.  Go ahead and bounce that playback and notice its not in sync...usually bounced early.

**Comment by steveschow** [comment]
ps - the first time I tried to use Jamstix free in Audiogridder (beta9), it played back correctly but then when I went to bounce it, something got fouled up where Jamstix was not responding to the transport at all.  I had to unload and reload Jamstix in order to refresh whatever so that transport would cause jamstix to go.  I'm not sure if this is somehow related to it being the "free" version.  UPDATE - it just happened again and this time when I went to make a change in Jamstix I got a message asking me to stop the host first.  The host was already stopped.  So I think something in AudioGridder is not updating the transport play/stop status and various related sync properly...  That's probably why sometimes it gets in a mode where I can't start it.

**Comment by steveschow** [comment]
my patience has run out.  Going back to trying to use VePro for jamstix for now.  I will check back in a few months to see if audiogridder handles it better.  Too much time wasted today trying to make it work in audiogridder, but there are quite a few jamstix users that would be very happy if audiogridder could work with Jamstix since Jamstix doesn't run properly on Mac anymore.

**Comment by steveschow** [comment]
beta10 is still out of sync for me.  Also, beta10 can't find audiogridder server running on parallels.

**Reply by apohl79** [reply]
I think it is a logic problem. Try the new manual delay control to adjust PDC. Thats how I got logic in sync. Reaper and Pro Tools were always in sync out of the box.

**Reply by apohl79** [reply]
I also tried just generating midi to feed another drum plugin. With PT and reaper it worked perfectly fine, with logic some manual pdc tweaks were needed.

**Reply by steveschow** [reply]
I also just tried DP11.  It is in better sync then LogicPro, but its still not sample accurate, the results bounce to audio 2ms late.  I tried both the AU and VST version of Audiogridder-inst

**Comment by steveschow** [comment]
also beta10 failed numerous plugin scans that used to work.

**Comment by Centripidity** [comment]
Thanks for beta10.  I'll download it and give it a test later and let you know how it goes.

Interesting that it performs differently in Logic, I wonder what's causing that.  Logic is certainly loaded with bugs so that might be the problem.  I know I frequently have to find work arounds for things that don't work the way they are supposed to.  At the moment I have a project where a particular group of MIDI regions randomly may or may not produce any sound when playing or bouncing.

Thanks for the adjustable PDC work around, that should help a lot.

I'm happy to do any extra testing you might need.  You can contact me at _peter at centripidity dot com_.

**Comment by dmuckala** [comment]
FYI, I've been running beta 6 with Logic to fix the delay comp problem and it works fine.  It didn't show up until beta 7.

**Reply by Centripidity** [reply]
Thanks, but I think this is a different issue because it also happens in beta6.  Jamstix is a little different from most other plugins as it receives no audio or MIDI data from Logic, just its transport information - primarily the number of beats since the start of the song.  It generates all of the audio from settings within Jamstix itself and that audio is streamed back to Logic.

**Comment by Centripidity** [comment]
Ok, I've done some testing with beta 10.  I'll post part one here and part two in a separate post.  Part two relates to the PDC you've added.

In general, the timing is much better but I'm still getting inconsistency between bounces.  I bounced a Jamstix track down in Logic seven times with a buffer of 23ms and the results are shown below.  The top four used Logic's offline, bounce in place function and the bottom three were real time bounces using the bounce button on the Stereo Out of the mixer.

The offline bounces showed latencies of between 3 and 15ms and the real time bounces between 15 and 24ms.

Better than beta9 but still not great.

<img width="1542" alt="Screen Shot 2022-07-07 at 16 12 15" src="https://user-images.githubusercontent.com/108377254/177710432-cff94f24-f5ae-4a2b-9109-f9a4cddd1ba6.png">
.

**Reply by apohl79** [reply]
cheers, thats good info!

**Comment by Centripidity** [comment]
Part 2

Next I tried using the new PDC to compensate and I did 11 offline bounces with a few different values of compensation from -23 to +12ms.  The results for the +12, +17 and +23ms are shown below.

The most striking (and important) result here is that, so far, the bounces seem to be consistent.  It appears that, on my system at least,  a compensation of between +12 and +17ms would be ideal.  Any chance of allowing values other than those in the menu for some future release?

If these results remain consistent (and it also works for multi-output streams) this will be very useable for my purposes.

I've been using the 16 x mono version of the plugin but so far have only been streaming a single stereo pair.

Thanks for this update and for being so responsive.


<img width="1542" alt="Screen Shot 2022-07-07 at 16 32 38" src="https://user-images.githubusercontent.com/108377254/177712209-7240c7a7-df03-44ca-8a5b-59d7b66a298f.png">

**Reply by apohl79** [reply]
Interesting indeed!

Regarding the PDC values, yes, that could be more customizable. 

When testing with logic, I was not able to find a consistent pattern by which the PDC samples have to be shifted to be in sync. With an 8 blocks buffer (1024 samples per block) I found -4 blocks manual delay to move things in sync with other instruments in the session. But for instance using a 4 blocks buffer with -2 blocks manual delay did not move it in sync. So something is going on..

**Reply by Centripidity** [reply]
Yeah, as per my post below.  I needed 12-17ms of PDC for a 23ms buffer but, with a multi-out instance, I only needed 6ms of PDC for a 186ms buffer.

**Reply by Centripidity** [reply]
What I find intriguing about this is that your PDC appears to be doing something that produces consistency while Logic is otherwise inconsistent.  It must be doing more than just changing the plugin latency reported to Logic I would have thought.

Something to think about perhaps?

**Comment by Centripidity** [comment]
I just tried Jamstix in multi-out mode.  At 23ms I was getting serious dropout (unsurprisingly given that I use a 128 sample buffer in Logic) so I bumped it up to 186ms.  Playback sounded good so I tried bouncing down all outputs.  With a compensation of 6ms it all looked spot on.  Consistent results from 3 attempts.

It's dinner time now (6pm) so I need to take a break.  It will be interesting to see if, when I close the project down and restart Logic (or create a new project), if the same compensation values will work.

This is looking very promising.

**Comment by noemirtil** [comment]
Hi all,
I too have delay compensations issues with AudioGridder, in Pro Tools 2020.12.
I couldn't exactly figure out how it works, but it appeared after one precise AG instance eating too much PTs resources on its own, don't know why. 
Seems also linked with the use of folder routing tracks inside a heavy mixing session. 
And when I try to freeze the guilty AG track then the delay compensation problem appears on some other folder routing track.
I'm actually stuck in my mix for this reason, and don't know how to find a workaround.
Any idea ?
